pallets/offchain-duster/src/lib.rs

```diff
@@ -0,0 +1,172 @@
+pub use pallet::*;
+
+pub const OFFCHAIN_WORKER_LOCK: &[u8] = b"hydradx/offchain-duster/lock/";
+pub const LOCK_DURATION: u64 = 100;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+	use frame_support::log;
+	use frame_support::pallet_prelude::*;
+	use frame_system::pallet_prelude::BlockNumberFor;
+
+	#[pallet::pallet]
+	pub struct Pallet<T>(_);
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T>
+	where
+		<T as orml_tokens::Config>::Balance: From<<T as pallet_duster::Config>::Balance>,
+		<T as pallet_duster::Config>::CurrencyId: From<<T as orml_tokens::Config>::CurrencyId>,
+	{
+		fn offchain_worker(block_number: <T as frame_system::Config>::BlockNumber) {
+			if let Err(e) = Self::_offchain_worker() {
+				log::info!(
+					target: "duster offchain worker",
+					"failed to run offchain worker at {:?}: {:?}",
+					block_number,
+					e,
+				);
+			} else {
+				log::debug!(
+					target: "duster offchain worker",
+					"offchain worker at block: {:?} completed!",
+					block_number,
+				);
+			}
+		}
+	}
+
+	#[pallet::config]
+	pub trait Config:
+		frame_system::Config + orml_tokens::Config + pallet_duster::Config + CreateSignedTransaction<DusterCall<Self>>
+	{
+		type AuthorityId: frame_system::offchain::AppCrypto<Self::Public, Self::Signature>;
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {}
+}
+impl<T: Config> Pallet<T> {
+
+	fn _offchain_worker() -> Result<(), OffchainErr>
+	where
+		<T as orml_tokens::Config>::Balance: From<<T as pallet_duster::Config>::Balance>,
+		<T as pallet_duster::Config>::CurrencyId: From<<T as orml_tokens::Config>::CurrencyId>,
+	{
+		if !sp_io::offchain::is_validator() {
+			return Err(OffchainErr::NotValidator);
+		}
+
+		let lock_expiration = Duration::from_millis(LOCK_DURATION);
+		let mut lock = StorageLock::<'_, Time>::with_deadline(&OFFCHAIN_WORKER_LOCK, lock_expiration);
+		let mut guard = lock.try_lock().map_err(|_| OffchainErr::OffchainLock)?;
+
+		for (who, asset, account) in orml_tokens::Accounts::<T>::iter() {
+			let ed = T::MinCurrencyDeposits::get(&asset.into());
+			let total = account.free.saturating_add(account.reserved);
+
+			if total < ed.into() {
+				let _ = Self::dust_account_signed(&who, asset.into());
+			}
+
+			guard.extend_lock().map_err(|_| OffchainErr::OffchainLock)?;
+		}
+
+		guard.forget();
+
+		Ok(())
+	}
+
+	fn dust_account_signed(
+		from: &T::AccountId,
+		currency_id: <T as pallet_duster::Config>::CurrencyId,
+	) -> Result<(), &'static str> {
+		let signer = Signer::<T, T::AuthorityId>::any_account();
+		if !signer.can_sign() {
+			return Err("No local accounts available. Consider adding one via `author_insertKey` RPC.");
+		}
+
+		let results = signer.send_signed_transaction(|_account| DusterCall::dust_account(from.clone(), currency_id));
+
+		for (acc, res) in &results {
+			match res {
+				Ok(()) => {
+					frame_support::log::info!("Dust moved successfully to [{:?}]", acc.id)
```

Comment:
We should refund the tx fee here

Reply:
hm. Not sure how we can do that here. I'll investigate.

There is a reward for successful account dusting. Isn't that enough ?  It is possible to configure it so it'd cover the tx fee + some additiional reward.